rlmolecule/mcts/mcts.py

```diff
@@ -0,0 +1,173 @@
+        # self.problem.store_policy_inputs_and_targets(state) # elided
+        children = vertex.children
+        if children is not None and len(children) > 0:
+            child = action_selection_function(vertex)
+            yield from self.run_from_vertex(child, action_selection_function, num_mcts_samples, mcts_selection_function)
+
+    def sample(
+            self,
+            vertex: MCTSVertex,
+            num_mcts_samples: int,
+            mcts_selection_function: Callable[[MCTSVertex], MCTSVertex],
+    ) -> None:
+        for _ in range(num_mcts_samples):
+            search_path, value = self._select(vertex, mcts_selection_function)
+            self._backpropagate(search_path, value)
+
+    def _select(
+            self,
+            root: MCTSVertex,
+            mcts_selection_function: Callable[[MCTSVertex], MCTSVertex],
+    ) -> ([MCTSVertex], float):
+        """
+        Selection step of MCTS
+        From Wikipedia (https://en.wikipedia.org/wiki/Monte_Carlo_tree_search):
+        Selection: Start from root R and select successive child vertices until a leaf vertex L is reached.
+        The root is the current game state and a leaf is any vertex that has a potential child from which no simulation
+        (playout) has yet been initiated. The section below says more about a way of biasing choice of child vertices that
+        lets the game tree expand towards the most promising moves, which is the essence of Monte Carlo tree search.
+        """
+        current = root
+        search_path = []
+        while True:
+            search_path.append(current)
+            children = current.children
+
+            if children is None:  # node is unexpanded: expand and return its value estimate
+                current.children = [self.get_vertex_for_state(state) for state in current.state.get_next_actions()]
+                return search_path, self._evaluate(current, search_path)
+
+            if len(children) == 0:  # node is expanded and terminal: return its value
+                return search_path, self.problem.get_reward(current.state)
+
+            current = mcts_selection_function(current)
+
+    def _evaluate(
+            self,
+            leaf: MCTSVertex,
+            search_path: [MCTSVertex],
+    ) -> float:
+        """
+        Expansion step of MCTS
+        From Wikipedia (https://en.wikipedia.org/wiki/Monte_Carlo_tree_search):
+        Expansion: Unless L ends the game decisively (e.g. win/loss/draw) for either player, create one (or more) child
+        vertices and choose vertex C from one of them. Child vertices are any valid moves from the game position defined by L.
+        """
+        children = leaf.children
+        state = leaf.state
+        if len(children) > 0:
+            child = random.choice(children)
+            search_path.append(child)
+            state = self._simulate(child.state)
+
+        return self.problem.get_reward(state)
```

Comment:
This is still a bit odd to me, I'm not sure why we don't just have the single evaluate function. In general, I like the idea of mirroring the four steps of MCTS as method names, but from the wikipedia they're actually the following, which I think makes more sense. 

1. select: more like the old tree_policy code; I'm not sure we need to care about selection function inside MCTS unless we think we'll change it often enough that subclassing would be a pain
2. expand: I liked having this be an explicit method, since it's different in MCTS vs. alphazero. In alphazero, steps 2 & 3 are tied together, but it doesn't mean we can't adopt and modify this structure for MCTS. It looks like you expand in `select`, where I think because of alphazero it makes more sense to check & expand in `simulation`. 
3. simulate: Here, it would combine _simulate and _evaluate, this was how I did it: https://github.com/NREL/rlmolecule/blob/884e9b00089e8e2f3eaaebdf89ad07ea534dbd9c/rlmolecule/mcts/mcts_node.py#L145-L162
4. backprogate: similar to how you have it is great